Case Study 1 - extra revisions (#91)
ImperialCollegeLondon/digital_research_sustainability_training: episodes/04_case_study1.md

She wants to assess the carbon emissions associated with both the development
and usage of her package and identify ways to reduce these emissions.

Celia should assess the balance of emissions involved in development of the code
base versus its usage. She should look at how to estimate these then focus her
emission reduction measures appropriately.

## Collecting Information

::::::::::::::::::::::::::::::::::::: challenge

### What data does Celia need to understand the emissions associated with her software?

- What are the key aspects of her work that Celia could estimate emissions for?
- What methodologies could Celia use to estimate her emissions?
- What additional data would she need to collect in each case?

::::::::::::::::::::::::::::::: solution

> 1. Use of laptop for development. Embodied emissions of the laptop should be readily

Reviewer: ```suggestion
1. **Use of laptop for software development**: Embodied emissions of the laptop should be readily
```

   findable in a PCF sheet. As her laptop underpins all of her work however it wouldn't
   be appropriate to ascribe the full embodied emissions to her package. She could
   ascribe a proportion of the embodied emissions to development of the package.
   Regarding the operational emissions she could attempt direct measurement of her
   laptop with a power meter but given the proportionally low operational emissions of
   consumer electronic devices a rough calculation using the Green Algorithms Calculator
   would likely be sufficient.

> 1. Use of LLMs. Tracking which model she uses, and how many queries she sends and the

Reviewer: ```suggestion
1. **Use of LLMs**: Tracking which model she uses, and how many queries she sends and the
```

   approximate size of the replies for use with the Hugging Face Ecologits calculator.

> 1. Execution of the package. Celia's software is used by many different users on a

Reviewer: ```suggestion
1. **Execution of the package**: Celia's software is used by many different users on a
```

   variety of systems. It's unlikely to be practical to get direct energy measurements
   in this case. More practically she could make use of the [Green Algorithms
   Calculator]. She should try and get as much detail as possible from users about how
   much they use the software and with what hardware. She's unlikely to be able to get
   information from all users but if she can get a representative sample then she could
   extrapolate to get a rough estimate. Estimating the embodied emissions for all of the
   machines the code is running on would be a potentially large job so she might want to
   put that out of scope.

> 1. Use of GitHub actions. Information about how often her workflows run and how long

Reviewer: ```suggestion
1. **Use of GitHub actions**: Information about how often her workflows run and how long
```

all of her workflows together have a runtime of around 2940 seconds. She adds ECO CI to
her workflow and runs notes its output over a few trial runs. The average of her trial
runs is around 1 gCO₂e for a workflow that runs for 500 seconds. This includes the
operational and embedded estimates.

### LLM Use

For creating inline documentation for her code, Celia has been using AI coding agents.
While she is not using them frequently, she notices that on an average, she writes
approximately 20 prompts to the agents every week. She takes note of the agent she uses

- GPT-5 mini - and that it typically provides short responses.

### Software Usage

Finally, Celia reaches out to her research group members who are users of her package.
They are able to provide her with the full specification of the machine they are using -
an [HP Z2 Tower G1i Workstation]. They run the package for around 18 hours a week using
all 20 cores.

[HP Z2 Tower G1i Workstation]: https://www.hp.com/gb-en/shop/product.aspx?id=8T229EA&opt=ABU&sel=DTP

From individuals she's in contact with, conversations she's had at conferences, mentions

> in academic papers and a workshop she ran recently Celia estimates that her code has

Reviewer: ```suggestion
in academic papers and a workshop she ran recently, Celia estimates that her code has
```

around 30 regular users outside of her own research group.

:::::::::::::::::::::::::::::::::::::::

## Analysis

::::::::::::::::::::::::::::::: challenge

## Estimating emissions

With the information provided in the previous section what estimates can you create for
Celia's emissions from different activities?

::::::::::::::::::::::: solution

- Software Development - From the model of her laptop she is able to find the PCF

>   datasheet from the manufacterer - [HP EliteBook 840 G9 PCF Sheet]. This gives a total

Reviewer: ```suggestion
  datasheet from the manufacturer - [HP EliteBook 840 G9 PCF Sheet]. This gives a total
```

  although it might make for a pretty rough estimate. This would give an estimate of
  around 17 kgCO₂e per week of operational emissions. Celia decides to leave out the
  embodied component of the analysis as she doesn't know enough about what hardware is
  being used to run her code.
- GitHub Actions - Given she has an estimate for a workflow of 500 seconds she chooses
  to simply scale this up to the full runtime of 1640 seconds. This gives an estimate of
  around 6 gCO₂e per week.
- LLM use - Using the HuggingFace EcoLogits calculator Celia estimates her weekly usage
  at around 1 gCO₂e.

[HP EliteBook 840 G9 PCF Sheet]: https://h20195.www2.hp.com/v2/GetDocument.aspx?docname=c09266068

:::::::::::::::::::::::

:::::::::::::::::::::::::::::::

## Taking Action

From the estimates Celia has made it's clear that the emissions associated with usage of
her package are the most significant. She also anticipates these growing over time given
the growing popularity of her package. The emissions from GitHub actions and LLM usage
are negligible.

:::::::::::::::::::::::::::::::::: challenge

## Measures to reduce emissions

> What measures can Celia take to redure to reduce the emissions from usage of her

Reviewer: ```suggestion
What measures can Celia take to reduce the emissions from the usage of her
```

- Making note the repetitive questions her package users might have and creating a
detailed user guide that includes instructions on how to make the most efficient use
of her package.

There are many steps Celia could take to reduce emissions. Keep a record of the ideas
you've had and compare them with those in the next section.

::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::: spoiler

## Celia Takes Action

Celia identifies several ways she can improve the emissions associated with usage
of her code base.

### Code Optimisation

> Optimize the code itself to make more effective use of available computational

Reviewer: ```suggestion
Optimise the code itself to make more effective use of available computational
```

resources. Celia uses a profiler with her code to identify areas where the code could be
optimised. She identifies the areas of the code where the bulk of the computation is
performed. After some experimentation she finds a way to improve use of SIMD in a key

> calculation. In her tests this gives a 5% performance boost to the code.

Reviewer: ```suggestion
calculation. Furthermore, she replaces the use of `pandas` with `polars` and reverses the order of a conditional statement and a loop deep into the code, so that the former is not checked several times unnecessarily. In her tests this gives a 7% performance boost to the code.
```

Reviewer: Suggestion based on Diego's comment: https://github.com/ImperialCollegeLondon/digital_research_sustainability_training/pull/85#discussion_r3128556814

with optimising the performance of the code. She provides them with some tips on how to
optimise the performance of the code when they run it on their local machines.
Additionally, she creates a detailed user guide that includes instructions on how to
make the most efficient use of her package, including tips on how to optimise the
performance of the code when running it on different hardware configurations.

Whilst it's difficult to estimate the overall impact of this work, with her help the
members of her research group that Celia works with were able to improve throughput when
using her package by 6%.

### Reducing Wasted Runs

Celia takes a pass at improving the error handling and input validation in her code to
reduce the likelihood of running into errors that lead to repeated runs of the code. She
implements a new configuration validation approach. She ways to catch some failure modes
early before significant computation has occured.

Again it's difficult to estimate the impact of this work but when these changes were
released Celia is contacted by several users confused by the new errors. This suggests
the changes are catching at least some errors.

> 

Reviewer: ```suggestion

### Hardware Usage

She decides to keep using the laptop for as long as its lifespan, instead of replacing it too soon. 

```

Reviewer: Suggestion based on Diego's comment: https://github.com/ImperialCollegeLondon/digital_research_sustainability_training/pull/85#discussion_r3128542111

### Other

Celia also integrates the [codecarbon](https://github.com/mlco2/codecarbon) as an
optional dependency in her code base so that it can report the carbon emissions when the
code is run. This allows her more easily to track the emissions associated with the
usage of her package.

:::::::::::::::::::::::::::::::::::::::::::

## Outcomes

<!-- markdownlint-disable-next-line line-length -->

> ![Carbon emissions for different research actions comparing pre- and post-intervention](fig/case_study1_outcomes.png){alt='A bar chart comparing the emissions from Software Development, GitHub Actions, LLM usage and Software usage before and after implementation of emissions reduction measures'}

Reviewer: I am missing where this plot is coming from.

Also the title says "Case Study 3 Outcomes" we might want to change it to Case Study 1?

Author: Thanks. Will update and add the script used to generate it.
